src/blob.rs

```diff
@@ -91,12 +95,39 @@ impl<'a> BlobObject<'a> {
         context: &'a Context,
         src: impl AsRef<Path>,
     ) -> std::result::Result<BlobObject<'a>, BlobError> {
-        let blob = BlobObject::create(context, src.as_ref().to_string_lossy(), b"")?;
-        fs::copy(src.as_ref(), blob.to_abs_path()).map_err(|err| {
-            fs::remove_file(blob.to_abs_path()).ok();
-            BlobError::new_copy_failure(blob.blobdir, &blob.name, src.as_ref(), err)
+        let src_file = fs::File::open(src.as_ref()).map_err(|err| {
+            BlobError::new_copy_failure(context.get_blobdir(), "", src.as_ref(), err)
+        })?;
+        let (stem, ext) = BlobObject::sanitise_name(&src.as_ref().to_string_lossy());
+        let (name, dst_file) = BlobObject::create_new_file(context.get_blobdir(), &stem, &ext)?;
+        BlobObject::copy_file(src_file, dst_file).map_err(|err| {
+            BlobError::new_copy_failure(context.get_blobdir(), &name, src.as_ref(), err)
         })?;
-        Ok(blob)
+        let blob = BlobObject {
+            blobdir: context.get_blobdir(),
+            name: format!("$BLOBDIR/{}", name),
+        };
+        context.call_cb(Event::NewBlobFile(blob.as_name().to_string()));
+        return Ok(blob);
+    }
+
+    // Copy a file in a source path to an already opened file.
+    //
+    // This reads the source file in 4Mb chunks.
+    fn copy_file(src_file: fs::File, dst_file: fs::File) -> Result<(), std::io::Error> {
+        let mut reader = BufReader::with_capacity(4 * 1024 * 1024, src_file);
+        let mut writer = BufWriter::new(&dst_file);
+        let mut buffer = reader.fill_buf()?;
+        let mut buf_len = buffer.len();
+        while buf_len > 0 {
```

Comment:
there is no need for this manual work io::copy already does all of this for you

Reply:
ooh, nice!  thanks for the pointer